backend/rates/history.go

```diff
@@ -0,0 +1,199 @@
+package rates
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"time"
+)
+
+// See the following for docs and details: https://www.coingecko.com/en/api.
+// Rate limit: 6k QPS per IP address.
+const coingeckoAPIV3 = "https://api.coingecko.com/api/v3"
+
+var (
+	// Copied from https://api.coingecko.com/api/v3/coins/list.
+	// The keyes must match entries in coins slice.
+	geckoCoin = map[string]string{
+		"BTC":  "bitcoin",
+		"LTC":  "litecoin",
+		"ETH":  "ethereum",
+		"USDT": "tether",
+		"LINK": "chainlink",
+		"MKR":  "maker",
+		"ZRX":  "0x",
+		"DAI":  "dai",
+		"BAT":  "basic-attention-token",
+		"USDC": "usd-coin",
+	}
+
+	// Copied from https://api.coingecko.com/api/v3/simple/supported_vs_currencies.
+	// The keys must match entries in fiats slice.
+	geckoFiat = map[string]string{
+		"USD": "usd",
+		"EUR": "eur",
+		"CHF": "chf",
+		"GBP": "gbp",
+		"JPY": "jpy",
+		"KRW": "krw",
+		"CNY": "cny",
+		"RUB": "rub",
+		"CAD": "cad",
+		"AUD": "aud",
+	}
+)
+
+// historyUpdateLoop periodically updates historical market exchange rates
+// forward in time starting with the last fetched timestamp.
+// It never returns.
+func (updater *RateUpdater) historyUpdateLoop() {
+	for {
+		// When to update next, after this loop iteration is done.
+		untilNext := 10 * time.Minute // TODO: add jitter
+
+		// TODO: Do this for all supported coins.
+		start := updater.historyLatestTimestamp("BTC")
+		if !start.IsZero() {
+			// Start slightly past the last fetched timestamp
+			start = start.Add(10 * time.Minute)
+			now := time.Now()
+			// It's ok if start == now or now < start. CoinGecko simply returns no results.
+			// It's also ok if our "now" doesn't match CoinGecko: it always returns
+			// the latest available data without errors.
+			err := updater.updateHistory("BTC", "USD", start, now)
+			if err != nil {
+				updater.log.Errorf("updateHistory(%v, %v): %v", start, now, err)
+				untilNext = time.Minute // TODO: exponential backoff
+			}
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+// backfillHistory fetches historical market exchange rates starting with
+// the earliest fetched timestamp backwards until the end of times.
+// TODO: what is "the end of times"?
+//
+// It does so in a loop and returns only after all data is backfilled.
+// Callers are expected to run this in a separate goroutine.
+func (updater *RateUpdater) backfillHistory() {
+	for {
+		// When to update next, after this loop iteration is done.
+		untilNext := 10 * time.Second // TODO: add jitter
+
+		// TODO: Do this for all supported coins.
+		end := updater.historyEarliestTimestamp("BTC")
+		if end.IsZero() {
+			end = time.Now()
+		}
+		// 90 days is the max interval which CoinGecko responds with hourly timeseries.
+		start := end.Add(-90 * 24 * time.Hour)
+		if time.Since(start) > 720*24*time.Hour {
+			// ~2 years of historical data is enough for now.
+			// TODO: Where do we stop?
```

Comment:
what about all of it, until coingecko does not have any more data?

Or stop at [2009-01-03](https://blockstream.info/block/000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f) :D

Reply:
Good point. :)
I've checked and the earlierst it returns is dated April 28, 2013 12:00:00 AM. If I give it a range like 2009-01-03 - 2010-01-03, it comes up with an empty list. So, let's fetch until we get an empty list.